bundles/org.openhab.binding.bluetooth/src/main/java/org/openhab/binding/bluetooth/AbstractBluetoothBridgeHandler.java

```diff
@@ -0,0 +1,223 @@
+    public void dispose() {
+        ScheduledFuture<?> inactiveRemovalJob = this.inactiveRemovalJob;
+        if (inactiveRemovalJob != null) {
+            inactiveRemovalJob.cancel(true);
+        }
+        this.inactiveRemovalJob = null;
+
+        synchronized (devices) {
+            for (BD device : devices.values()) {
+                removeDevice(device);
+            }
+        }
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+    }
+
+    private void removeInactiveDevices() {
+        // clean up orphaned entries
+        synchronized (devices) {
+            for (BD device : devices.values()) {
+                if (shouldRemove(device)) {
+                    logger.debug("Removing device '{}' due to inactivity", device.getAddress());
+                    removeDevice(device);
+                }
+            }
+        }
+    }
+
+    protected void removeDevice(BluetoothDevice device) {
+        device.dispose();
+        synchronized (devices) {
+            devices.remove(device.getAddress());
+        }
+        discoveryListeners.forEach(listener -> listener.deviceRemoved(device));
+    }
+
+    private boolean shouldRemove(BluetoothDevice device) {
+        // we can't remove devices with listeners since that means they have a handler.
+        if (device.hasListeners()) {
+            return false;
+        }
+        // devices that are connected won't receive any scan notifications so we can't remove them for being idle
+        if (device.getConnectionState() == ConnectionState.CONNECTED) {
+            return false;
+        }
+
+        // we remove devices we haven't seen in a while
+        return ZonedDateTime.now().minusSeconds(config.inactiveDeviceCleanupThresholdSecs)
+                .isAfter(device.getLastSeenTime());
+    }
+
+    @Override
+    public void addDiscoveryListener(BluetoothDiscoveryListener listener) {
+        discoveryListeners.add(listener);
+    }
+
+    @Override
+    public void removeDiscoveryListener(@Nullable BluetoothDiscoveryListener listener) {
+        discoveryListeners.remove(listener);
+    }
+
+    @Override
+    public void scanStart() {
+        // Enable scanning even while discovery is disabled in config. This allows manual starting discovery.
+        activeScanEnabled = true;
+        refreshDiscoveredDevices();
+    }
+
+    protected void refreshDiscoveredDevices() {
+        logger.debug("Refreshing Bluetooth device list...");
+        synchronized (devices) {
+            devices.values().forEach(this::deviceDiscovered);
+        }
+    }
+
+    @Override
+    public void scanStop() {
+        // Set active discovery state back to the configured discovery state.
+        activeScanEnabled = false;
+        // We need to keep the adapter in discovery mode as we otherwise won't get any RSSI updates either
+    }
+
+    @Override
+    public BD getDevice(BluetoothAddress address) {
+        synchronized (devices) {
```

Comment:
why not use `computeIfAbsent` here? `ConcurrentHashMap` implements that atomically.

Reply:
I toyed with the idea, but the atomic behavior is already handled by the synchronized blocks and computeIfAbsent saves at most 2 lines of code while at the same having to utilize the overhead of a multi-statement lambda. I typically use computeIfAbsent if the lambda can fit on a single line.

So to answer your question: I didn't implement it for stylistic reasons.

But you did bring up a good point the I'm using a `ConcurrentHashMap` which is overkill in this particular case, so I'll change it to a `HashMap`.